pkg/addonmanager/controllers/agentdeploy/hosted_sync.go

```diff
@@ -0,0 +1,143 @@
+	}
+
+	installMode, hostingClusterName := constants.GetHostedModeInfo(addon.GetAnnotations())
+	if installMode != constants.InstallModeHosted {
+		// the installMode is changed from hosted to default, cleanup the hosting resources
+		if err := s.cleanupDeployWork(ctx, addon, ""); err != nil {
+			return addon, err
+		}
+		addonRemoveFinalizer(addon, constants.HostingManifestFinalizer)
+		return addon, nil
+	}
+
+	// Get Hosting Cluster, check whether the hosting cluster is a managed cluster of the hub
+	// TODO: check whether the hosting cluster of the addon is the same hosting cluster of the klusterlet
+	hostingCluster, err := s.controller.managedClusterLister.Get(hostingClusterName)
+	if errors.IsNotFound(err) {
+		if err = s.cleanupDeployWork(ctx, addon, ""); err != nil {
+			return addon, err
+		}
+
+		meta.SetStatusCondition(&addon.Status.Conditions, metav1.Condition{
+			Type:    constants.HostingClusterValidity,
+			Status:  metav1.ConditionFalse,
+			Reason:  constants.HostingClusterValidityReasonInvalid,
+			Message: fmt.Sprintf("hosting cluster %s is not a managed cluster of the hub", hostingClusterName),
+		})
+
+		addonRemoveFinalizer(addon, constants.HostingManifestFinalizer)
+		return addon, nil
+	}
+	if err != nil {
+		return addon, err
+	}
+	meta.SetStatusCondition(&addon.Status.Conditions, metav1.Condition{
+		Type:    constants.HostingClusterValidity,
+		Status:  metav1.ConditionTrue,
+		Reason:  constants.HostingClusterValidityReasonValid,
+		Message: fmt.Sprintf("hosting cluster %s is a managed cluster of the hub", hostingClusterName),
+	})
+
+	if !hostingCluster.DeletionTimestamp.IsZero() {
+		if err = s.cleanupDeployWork(ctx, addon, hostingClusterName); err != nil {
+			return addon, err
+		}
+		addonRemoveFinalizer(addon, constants.HostingManifestFinalizer)
+		return addon, nil
+	}
+
+	if !addon.DeletionTimestamp.IsZero() {
+		// clean up the deploy work until the hook work is completed
+		if addonHasFinalizer(addon, constants.HostingPreDeleteHookFinalizer) {
+			return addon, nil
+		}
+
+		if err = s.cleanupDeployWork(ctx, addon, hostingClusterName); err != nil {
+			return addon, err
+		}
+		addonRemoveFinalizer(addon, constants.HostingManifestFinalizer)
+		return addon, nil
+	}
+
+	if addonAddFinalizer(addon, constants.HostingManifestFinalizer) {
+		return addon, nil
+	}
+
+	deployWorkName := constants.DeployHostingWorkName(addon.Namespace, addon.Name)
+	deployWork, _, err := s.controller.buildManifestWorks(ctx, s.agentAddon, installMode, hostingClusterName, cluster, addon)
+	if err != nil {
+		return addon, err
+	}
+
+	if deployWork == nil {
+		if err = s.cleanupDeployWork(ctx, addon, hostingClusterName); err != nil {
+			return addon, fmt.Errorf("failed to cleanup deploy work %s/%s when got no object", hostingClusterName, deployWorkName)
+		}
+
+		meta.SetStatusCondition(&addon.Status.Conditions, metav1.Condition{
+			Type:    constants.AddonHostingManifestApplied,
+			Status:  metav1.ConditionFalse,
```

Comment:
this also seems weird. we should either set it to true or do not set it.

Reply:
removed the SetStatusCondition